qemu: enable some basic tests

risicle opened the pull request:
###### Motivation for this change
Give us some extra certainty about the packages we're producing & allow breakages to be caught before a full `nixpkgs-review`.

Patching is deliberately performed in the `preCheck`  to avoid adding `socat` as a dependency if the tests are disabled.

There are plenty more tests for qemu, but most of those are significantly harder to get working, especially in a restricted build environment. This is better than nothing.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

risicle commented:
lol

```
Unexpected error in unix_listen_saddr() at ../util/qemu-sockets.c:933:
UNIX socket path '/private/tmp/nix-build-qemu-host-cpu-only-for-vm-tests-6.1.0.drv-0/qemu-test_acpi_q35_tcg_tis.OV8RC1/sock' is too long
```

:scratch-head: That's going to be a fun one to work around

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -217,6 +216,32 @@ stdenv.mkDerivation rec {
+    substituteInPlace ../tests/unit/meson.build \
+      --replace 'timeout: slow_tests' 'timeout: 50 * slow_tests'
+    substituteInPlace ../tests/qtest/meson.build \
+      --replace 'timeout: slow_qtests' 'timeout: 50 * slow_qtests'
```
Might be worth seeing if upstream will increase their timeouts — I imagine fewer spurious failures due to timeouts would make development easier.

risicle replied:
Might do yes.

alyssais reviewed (approved):
This looks great!  What's still to do?

risicle commented:
I'm currently keeping this on the shelf until we've got 21.11 out the door. I think the next thing I need to look at is how to handle the unix socket path length problem that arises when the package's name is too long, i.e. the `qemu-host-cpu-only-for-vm-tests-6.1.0` variant.

risicle committed:
qemu: fixup basic tests, enable in passthru.tests

these may not be reliable enough to enable by default, but enabling
them as a passthru may allow us to get a feel for which platforms
have trouble with them

risicle commented:
Ok, this is now happy for me on x86 linux and darwin.

risicle commented:
@alyssais it's been a while - would you like to re-review? I'm now only considering enabling the tests in a `passthru.tests` variant to start with.

alyssais reviewed (approved):
LGTM.

It would still be nice if we could propose the timeout increases upstream at some point so we could avoid modifications here.
